### PR TITLE
repair: Track repair_meta created on both repair follower and master

### LINKED ISSUE #7009: Track repair_meta created on both repair follower and master
It is pretty hard to find the repair_meta object when debugging a core.
We can put the repair_meta object created by both repair follower and master into a map.

### PR DESCRIPTION
**Pull request by asias**
It is pretty hard to find the repair_meta object when debugging a core.
This patch makes it is easier by putting repair_meta object created by
both repair follower and master into a map.

Fixes #7009

### CODE REVIEW

**Comment by denesb** [comment]
I think better yet would be to just link them into an intrusive list in the debug namespace.

**Comment by denesb** [comment]
Like we do with seastar threads and sstables.

File: repair/row_level.cc [hunk 1]
```diff
@@ -2523,7 +2534,7 @@ class row_level_repair {
             auto schema_version = s->version();
             bool table_dropped = false;
 
-            repair_meta master(_ri.db,
+            auto rm = make_lw_shared<repair_meta>(_ri.db,
```
Comment (avikivity) [inline repair/row_level.cc:2537]: Why this change? You could keep it as it is, and register a pointer to `master` in the map.

File: repair/row_level.cc [hunk 2]
```diff
@@ -1844,6 +1843,18 @@ class repair_meta {
...
+std::unordered_map<node_repair_meta_id, lw_shared_ptr<repair_meta>>& repair_meta::repair_meta_map() {
+    return _repair_metas;
+}
+
+// The repair_metas created actively, i.e., local node is the repair master.
+static thread_local std::unordered_map<node_repair_meta_id, lw_shared_ptr<repair_meta>> _repair_metas_masters;
+std::unordered_map<node_repair_meta_id, lw_shared_ptr<repair_meta>>& repair_meta::repair_meta_map_masters() {
+    return _repair_metas_masters;
+}
+
```
Comment (avikivity) [inline repair/row_level.cc:1857]: We often keep such debug information in the debug namespace, consider doing the same.

Comment (denesb) [reply repair/row_level.cc:1857]: Also note that using a map has no advantage while debugging, we don't have the hash or the less functions available (or runnable) so the way to find anything in them is to iterate over all items. You might as well just use a simple `std::list` or `boost::intrusive::list`.

File: repair/row_level.cc (hunk 1 above)
Comment (asias) [reply repair/row_level.cc:2537]: I wanted the _repair_metas_masters  and _repair_metas (existing) one to be the same.

Comment (asias) [reply repair/row_level.cc:2537]: Using row pointer is even better in turns of help debugging.

File: repair/row_level.cc (hunk 2 above)
Comment (asias) [reply repair/row_level.cc:1857]: This is a good point.  Using list is simpler but it is slower to remove the items. We are not creating those objects on the fast path. I think it is fine.

**Comment by asias** [comment]
> I think better yet would be to just link them into an intrusive list in the debug namespace.

Can you point me the relevant code?

File: repair/row_level.cc [hunk 3]
```diff
@@ -764,10 +764,9 @@ class repair_meta {
         });
     }
 
-    static std::unordered_map<node_repair_meta_id, lw_shared_ptr<repair_meta>>& repair_meta_map() {
-        static thread_local std::unordered_map<node_repair_meta_id, lw_shared_ptr<repair_meta>> _repair_metas;
```
Comment (denesb) [inline repair/row_level.cc:768]: Very happy to see this static move out the method, working with function statics is a major pain in gdb. It has a very weird syntax that is so easy to get wrong (or forget) and gdb often forgets the type, and spelling out correctly a template (with all the hidden defaults) is one of the hardest things to do in gdb.

Comment (asias) [reply repair/row_level.cc:768]: Yes. That is why I am moving this out. It is painful to work with it!

Comment (asias) [reply repair/row_level.cc:768]: btw, sometimes, thread_local is not working. so, I am considering do not use the thread_local for the repair_meta objects in debug space.

**Comment by denesb** [comment]
> > I think better yet would be to just link them into an intrusive list in the debug namespace.
> 
> Can you point me the relevant code?

https://github.com/scylladb/scylla/blob/74db08165ddcc2db061b88b4a28e0fa682305655/sstables/sstables.hh#L135
https://github.com/scylladb/scylla/blob/74db08165ddcc2db061b88b4a28e0fa682305655/sstables/sstables.hh#L545
https://github.com/scylladb/scylla/blob/74db08165ddcc2db061b88b4a28e0fa682305655/sstables/sstables.cc#L93-L105
https://github.com/scylladb/scylla/blob/74db08165ddcc2db061b88b4a28e0fa682305655/sstables/sstables.cc#L3512-L3514
Note that `bi::constant_time_size<false>` parameter to `bi::list`, causing the members to be automatically removed on destruction.

File: repair/row_level.cc (hunk 2 above)
Comment (denesb) [reply repair/row_level.cc:1857]: Removing is updating 2 pointers, I don't think it gets much cheaper than that. :)

File: repair/row_level.cc (hunk 3 above)
Comment (denesb) [reply repair/row_level.cc:768]: How do you plan on avoiding `thread_local`? We can't have all threads use the same list, that will result in corruption.

I think you can rely on `thread_local` working, it is such a fundamental building block in our code that without thread debugging there is really no debugging at all.

File: repair/row_level.cc (hunk 2 above)
Comment (asias) [reply repair/row_level.cc:1857]: We need to search the repair_meta belongs to a specific node_repair_meta_id when we remove it.

File: repair/row_level.cc (hunk 3 above)
Comment (asias) [reply repair/row_level.cc:768]: std::vector<std::list<*repair_meta>> metas, metas[x] can only be accessed on shard x.

**Comment by asias** [comment]
Changes in 238cefd8d740d5d459f10e4de3c84f946c296f3e:
- Use boost intrusive list to track repair_meta objects
- Drop _repair_metas_masters
- Export repair_meta list in debug namespace

File: repair/row_level.cc [hunk 4]
```diff
@@ -2686,3 +2720,10 @@ future<> repair_service::stop() {
 repair_service::~repair_service() {
     assert(_stopped);
 }
+
+namespace debug {
+namespace repair {
+    static thread_local repair_meta_tracker* repair_meta_for_masters = &::_repair_meta_for_masters;
+    static thread_local repair_meta_tracker* repair_meta_for_followers = &::_repair_meta_for_followers;
+}
```
Comment (denesb) [inline repair/row_level.cc:2728]: I think these pointers are unnecessary, you can just wrap the `_repair_meta_for_masters` and `_repair_meta_for_followers` in `namespace debug::repair { }` and it will be enough.

Comment (asias) [reply repair/row_level.cc:2728]: I wanted to avoid forward declarations for  _repair_meta_for_masters and _repair_meta_for_followers. Yes, I think it is better to put them inside debug namespace.

**Comment by asias** [comment]
Changes in b280f67764f1a5f75199cfb0fffeda89ff9a7d88:
- Drop the point wrapper in debug name space.

File: repair/row_level.cc (hunk 4 above)
Comment (asias) [reply repair/row_level.cc]: I've sent a new version.

File: repair/row_level.cc [hunk 5]
```diff
@@ -2235,6 +2250,28 @@ future<> repair_uninit_messaging_service_handler() {
...
+public:
+    void add(repair_meta& rm) {
+        _repair_metas.push_back(rm);
+    }
+};
+
+namespace debug {
+    extern thread_local repair_meta_tracker repair_meta_for_masters;
+    extern thread_local repair_meta_tracker repair_meta_for_followers;
+}
```
Comment (denesb) [inline repair/row_level.cc:2266]: Why do we need separate forward declaration and defninition later? It's in the same file, can't we just define here and be done with it?

Comment (asias) [reply repair/row_level.cc:2266]: Nice! Yes, we can do that. I was fooled by class dependency thing.

**Comment by asias** [comment]
Changes in e9a520a22b8555056c77050d76c412694cc2e619:
- Drop the unnecessary forward declaration.

**Comment by avikivity** [comment]
Looks fine, will wait for @denesb ack.

**Comment by denesb** [comment]
LGTM.
